### PR TITLE
Add libs/auth tests, part 3

### PR DESCRIPTION
**Pull request by arsalansufi**
_Easiest reviewed by commit_

## Overview

Issue link: https://github.com/votingworks/vxsuite/issues/3057

More `libs/auth` tests, this time for the top-level auth API implementations

## Checklist

- [ ] ~I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced~ N/A
- [x] I have added JSDoc comments to any newly introduced exports

### CODE REVIEW

File: libs/types/src/auth/auth.ts [hunk 1]
```diff
@@ -28,6 +28,8 @@ export type User =
   | PollWorkerUser
   | CardlessVoterUser;
 
+export type UserWithCard = Exclude<User, CardlessVoterUser>;
```
[the remark this answers is not in the document]
Comment (jonahkagan) [reply libs/types/src/auth/auth.ts:31]: Great idea

File: libs/auth/test/utils.ts [hunk 2]
```diff
@@ -10,17 +12,47 @@ export function numericArray(input: {
...
+export function mockCardAssertComplete(mockCard: MockCard): void {
+  for (const method of [
+    'getCardStatus',
+    'checkPin',
+    'program',
+    'readData',
+    'writeData',
+    'clearData',
+    'unprogram',
+  ] as const) {
```
Comment (jonahkagan) [inline libs/auth/test/utils.ts:55]: Any way to make sure this doesn't get out of date if we add a method?

File: libs/auth/src/inserted_smart_card_auth.test.ts [hunk 3]
```diff
@@ -1,20 +1,928 @@
...
+
+  mockCardStatus({ status: 'ready', user });
+  expect(await auth.getAuthStatus(machineState)).toEqual({
+    status: 'checking_pin',
+    user,
+  });
+
+  mockCard.checkPin
+    .expectCallWith(wrongPin)
+    .resolves({ response: 'incorrect', numRemainingAttempts: Infinity });
```
Comment (jonahkagan) [inline libs/auth/src/inserted_smart_card_auth.test.ts:225]: Do we need any testing for what happens when there are no remaining pin attempts and the user tries to enter their pin? I can't remember what happens in that case and what code is responsible for it

**Review by jonahkagan — APPROVED** [review]
Exhaustive! Exhausting! 😅

File: libs/auth/src/inserted_smart_card_auth.test.ts (hunk 3 above)
Comment (arsalansufi) [reply libs/auth/src/inserted_smart_card_auth.test.ts:225]: Yeah good q! I haven't actually implemented that functionality yet. Will be doing so as part of https://github.com/votingworks/vxsuite/issues/2527

File: libs/auth/test/utils.ts (hunk 2 above)
Comment (arsalansufi) [reply libs/auth/test/utils.ts]: Good call! Came up with this: https://github.com/votingworks/vxsuite/pull/3145/commits/20cc3e4648e4404eb05344c3f31344986e9d6e3e

Comment (arsalansufi) [reply libs/auth/test/utils.ts]: ^ (TS will complain if we're not exhaustive)

example

Comment (jonahkagan) [reply libs/auth/test/utils.ts]: Nice, that's clever.

I looked to see how we do it in the Grout client and [it's a bit simpler](https://github.com/votingworks/vxsuite/blob/7ca8e5abd1953b51f8daf71e04be4e7e8b8385ed/libs/grout/test-utils/src/mock_client.ts#L70-L74). Wondering if the same approach would work here

Comment (arsalansufi) [reply libs/auth/test/utils.ts]: Ah nice, yeah much simpler: https://github.com/votingworks/vxsuite/pull/3145/commits/94ff53443e3197b6fb30700c0beff85b82e1adc0

**Pull request merged**